core/src/main/java/com/linecorp/armeria/common/ExceptionSampler.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Map;
+import java.util.function.Function;
+
+import org.jctools.maps.NonBlockingHashMap;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.util.Sampler;
+
+final class ExceptionSampler implements Sampler<Class<? extends Throwable>> {
+
+    private final Map<Class<? extends Throwable>, Sampler<Class<? extends Throwable>>> samplers =
+            new NonBlockingHashMap<>();
+    private final Function<? super Class<? extends Throwable>,
+            ? extends Sampler<Class<? extends Throwable>>> samplerFactory;
+    private final String spec;
+
+    ExceptionSampler(String spec) {
+        samplerFactory = unused -> Sampler.of(spec);
+        this.spec = spec;
+    }
+
+    @Override
+    public boolean isSampled(Class<? extends Throwable> exceptionType) {
```

Comment:
so the rationale for the type parameter is solving the overhead problem caused by allocating an exception, correct?

Reply:
Yeah, kind of. If we accepted a `Throwable`, we will see a chicken and egg problem. :smile: